File: blockchain-info-aggregation-stream-parent/blockchain-info-aggregation-stream-infrastructure/src/main/java/info/blockchain/aggregation/infrastructure/stream/ReplaySubjectMessageStream.java
package info.blockchain.aggregation.infrastructure.stream;

import reactor.core.publisher.DirectProcessor;
import reactor.core.publisher.FluxProcessor;

import java.util.function.Function;

public class ReplaySubjectMessageStream<T> implements MessageStream<T> {

    private final FluxProcessor<T, T> subject = DirectProcessor.create();

    @Override
    public MessagePublisher<T> getPublisher() {
        return obj -> {
            subject.onNext(obj);
            return obj;
        };
    }

    @Override
    public MessageSubscriber<T> getSubscriber() {
        return () -> subject.map(Function.identity());
    }

}

File: blockchain-info-aggregation-stream-parent/blockchain-info-aggregation-stream-infrastructure/src/main/java/info/blockchain/aggregation/infrastructure/stream/ReplaySubjectMessageStream.kt
package info.blockchain.aggregation.infrastructure.stream

import org.reactivestreams.Publisher
import reactor.core.publisher.DirectProcessor

class ReplaySubjectMessageStream<T> : MessageStream<T> {

    private val subject = DirectProcessor.create<T>()

    override val publisher: MessagePublisher<T>
        get() = object : MessagePublisher<T> {
            override fun publish(obj: T): T {
                subject.onNext(obj)
                return obj
            }
        }

    override val subscriber: MessageSubscriber<T>
        get() = object : MessageSubscriber<T> {
            override fun publisher(): Publisher<T> {
                return subject.map { x -> x }
            }
        }


}
